video/out/wayland_common.c

```diff
@@ -1297,3 +1313,31 @@ bool vo_wayland_config (struct vo *vo, uint32_t flags)
 
     return true;
 }
+
+void vo_wayland_request_frame(struct vo *vo, void *data, vo_wayland_frame_cb cb)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+    wl->frame.data = data;
+    wl->frame.function = cb;
+    MP_DBG(wl, "restart frame callback\n");
+    frame_callback(wl, NULL, 0);
+}
+
+bool vo_wayland_wait_frame(struct vo *vo)
+{
+    struct vo_wayland_state *wl = vo->wayland;
+
+    if (!wl->frame.callback)
+        return false;
+
+    // Reduce timeout if there are more than 3 pending waits.
+    // Most likely the mpv isn't receiving frame callbacks then.
+    // So don't timeout too long to avoid AV desync.
+    // 10ms matches the time specified in eglSwapInterval(0) mesa patch for wayland.
+    const int timeout = (wl->frame.pending > 3 ? 10 : 100); // ms
```

Comment:
>If you block and never timeout, mpv will get stuck if it's not visible (not receiving frame callbacks)

I think we can reasonably assume that the compositor renders mpv at a relatively high framerate (say, more than 10 frames a second) if mpv is visible. This would mean that if after 100ms no new frame callback arrives, we could assume it's invisible and drop. Otherwise, we keep mpv blocked. This should take care of it.